### PR TITLE
[Feat] support teacache for WAN2.2

### PR DESCRIPTION
**Pull request by RuixiangMa**
## Purpose
support teacache for WAN2.2

## Test Result

GPU: 2* 4090（24G) 

Modify image_to_video.py to support TeaCache

python image_to_video.py   --model Wan-AI/Wan2.2-TI2V-5B-Diffusers   --image cherry_blossom.jpg   --prompt "Cherry blossoms swaying gently in the breeze, petals falling, smooth motion"   --negative-prompt "blurry, distorted, poor quality, artifacts"   --height 480   --width 832   --num-frames 48   --guidance-scale 4.0   --num-inference-steps 40   --flow-shift 12.0   --fps 16 --cache-backend tea_cache  --tensor-parallel-size 2 --enable-cpu-offload --output i2v_output02.mp4

| Metric   | NO TeaCache | TeaCache |
|----------|-------------|----------|
| Video    | <video controls width="320" src="https://github.com/user-attachments/assets/b8919dc4-81b1-49ea-9774-a5b8d59445e9"></video> | <video controls width="320" src="https://github.com/user-attachments/assets/5d61e236-30c9-4ac9-81ba-a3bd84230f20"></video> |
| Time     | 46.645  s    | 25.331s |

python examples/offline_inference/text_to_video/text_to_video.py  --model /data/models/Wan-AI/Wan2.2-TI2V-5B-Diffusers  --prompt "Two anthropomorphic cats in comfy boxing gear and bright gloves fight intensely on a spotlighted stage."   --negative-prompt "色调艳丽，过曝，静态，细节模糊不清，字幕，风格，作品，画作，画面，静止，整体发灰，最差质量，低质量，JPEG压缩残留，丑陋的，残缺的，多余的手指，画得不好的手部，画得不好的脸部，畸形的，毁容的，形态畸形的肢体，手指融合，静止动的画面，杂乱的背景，三条腿，背景人很多，倒着走"   --height 480   --width 832   --num-frames 33   --guidance-scale 4.0   --guidance-scale-high 4.0   --boundary-ratio 0.875   --num-inference-steps 40   --flow-shift 12.0   --fps 16 --enable-cpu-offload  --tensor-parallel-size 2 --output t2v_out.mp4

| Metric   | NO TeaCache | TeaCache |
|----------|-------------|----------|
| Video    | <video controls width="320" src="https://github.com/user-attachments/assets/89341f96-bb93-4dfd-b2ef-f1e31fdf969f"></video> | <video controls width="320" src="https://github.com/user-attachments/assets/865bea60-9c3d-4e2e-88aa-187f2d0ebc6a"></video> |
| Time     | 31.8797  s    | 17.562s |

### CODE REVIEW

**Review by lishunyang12 — COMMENTED** [review]
Great addition! The 2x speedup on 480x832 is impressive. The extractor follows the established pattern nicely, which makes the code easy to follow.

Left a couple of comments below.

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 1]
```diff
@@ -380,6 +380,119 @@ def postprocess(h):
...
+        ).chunk(6, dim=2)
+        shift_msa = shift_msa.squeeze(2)
+        scale_msa = scale_msa.squeeze(2)
+        c_shift_msa = c_shift_msa.squeeze(2)
+        c_scale_msa = c_scale_msa.squeeze(2)
+    else:
+        shift_msa, scale_msa, _, c_shift_msa, c_scale_msa, _ = (
+            block.scale_shift_table + timestep_proj.float()
+        ).chunk(6, dim=1)
+
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py:469]: Add a comment explaining when each branch is hit (e.g. `# 4D: I2V/TI2V path with per-token timestep embedding` vs `# 2D: T2V path`) — the 4D case is the I2V/TI2V path with an extra sequence dimension.

File: vllm_omni/diffusion/cache/teacache/config.py [hunk 2]
```diff
@@ -37,6 +37,15 @@
         3.20000000e00,
         -2.00000000e-02,
     ],
+    # Default WanTransformer3DModel coefficients for 480x832 resolution
+    # the other resolution could need to be tuned specifically
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/config.py:58]: The coefficients are resolution-dependent and this was tested only on 480x832. Add a log warning when TeaCache is enabled for Wan2.2 at a different resolution, so users know to expect quality differences.

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 3]
```diff
@@ -380,6 +380,119 @@ def postprocess(h):
...
+    module: nn.Module,
+    hidden_states: torch.Tensor,
+    timestep: torch.LongTensor,
+    encoder_hidden_states: torch.Tensor,
+    encoder_hidden_states_image: torch.Tensor | None = None,
+    return_dict: bool = True,
+    attention_kwargs: dict[str, Any] | None = None,
+    **kwargs: Any,
+) -> CacheContext:
+    """
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py:404]: `encoder_hidden_states_image` is passed straight to `module.condition_embedder(...)` with no None check. `WanTransformer3DModel` handles `None` internally so it's fine, but add an inline comment noting it's `None` for text-to-video and only set for image-to-video.

**Review by lishunyang12 — COMMENTED** [review]
Solid extractor implementation that follows the existing patterns well. A few things worth addressing below.

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 4]
```diff
@@ -380,6 +380,119 @@ def postprocess(h):
...
+        encoder_hidden_states = torch.concat([encoder_hidden_states_image, encoder_hidden_states], dim=1)
+
+    block = module.blocks[0]
+    if timestep_proj.ndim == 4:
+        shift_msa, scale_msa, _, c_shift_msa, c_scale_msa, _ = (
+            block.scale_shift_table.unsqueeze(0) + timestep_proj.float()
+        ).chunk(6, dim=2)
+        shift_msa = shift_msa.squeeze(2)
+        scale_msa = scale_msa.squeeze(2)
+        c_shift_msa = c_shift_msa.squeeze(2)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py]: `c_shift_msa` and `c_scale_msa` are computed in both branches but never referenced afterward. Ignore those chunk outputs with `_` like you already do for the gate values.

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 5]
```diff
@@ -380,6 +380,119 @@ def postprocess(h):
     )
 
 
+def extract_wan2_2_context(
+    module: nn.Module,
+    hidden_states: torch.Tensor,
+    timestep: torch.LongTensor,
+    encoder_hidden_states: torch.Tensor,
+    encoder_hidden_states_image: torch.Tensor | None = None,
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py:399]: `attention_kwargs` is accepted but never forwarded to `run_transformer_blocks` or anywhere else. Intentional for signature compatibility, or should it be threaded through to the block calls?

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 6]
```diff
@@ -380,6 +380,119 @@ def postprocess(h):
...
+            h = block(h, encoder_hidden_states, timestep_proj, rotary_emb)
+        return (h,)
+
+    def postprocess(h):
+        shift, scale = module.output_scale_shift_prepare(temb)
+        shift = shift.to(h.device)
+        scale = scale.to(h.device)
+        if shift.ndim == 2:
+            shift = shift.unsqueeze(1)
+            scale = scale.unsqueeze(1)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py:483]: The explicit `.to(h.device)` calls for shift and scale stand out — no other extractor (Qwen, Bagel, ZImage) does this in postprocess. Did you hit an actual device mismatch? If it's needed, add a short comment explaining why.

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 7]
```diff
@@ -380,6 +380,119 @@ def postprocess(h):
...
+    else:
+        shift_msa, scale_msa, _, c_shift_msa, c_scale_msa, _ = (
+            block.scale_shift_table + timestep_proj.float()
+        ).chunk(6, dim=1)
+
+    norm_hidden_states = (block.norm1(hidden_states.float()) * (1 + scale_msa) + shift_msa).type_as(hidden_states)
+    modulated_input = norm_hidden_states
+
+    def run_transformer_blocks():
+        h = hidden_states
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py:474]: `timestep_proj_prepare` can return a sharded tensor while `hidden_states` is still the full sequence here (before the blocks). Did you test this with multi-GPU sequence parallelism, or only single-GPU? If SP isn't supported yet, note that explicitly.

File: vllm_omni/diffusion/cache/teacache/config.py (hunk 2 above)
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/config.py:58]: How were these coefficients derived? The other entries cite their source (TeaCache paper, ComfyUI-TeaCache empirical tuning) — add a similar note here.

**Comment by Gaohan123** [comment]
@RuixiangMa Hello, any updates?

**Comment by RuixiangMa** [comment]
> @RuixiangMa Hello, any updates?

I'll process the comments today

File: vllm_omni/diffusion/cache/teacache/config.py (hunk 2 above)
Comment (RuixiangMa) [reply vllm_omni/diffusion/cache/teacache/config.py:58]: Just referenced the official Wan2.1 [coefficients](https://github.com/ali-vilab/TeaCache).

Comment (RuixiangMa) [reply vllm_omni/diffusion/cache/teacache/config.py:58]: I only tested 480p because it ran oom（only avaliable gpu（24G))

**Comment by RuixiangMa** [comment]
@lishunyang12 Sorry for the late reply，all comments are resolved. Also fixed a Teacache and CPU offload incompatibility bug PTAL

File: vllm_omni/diffusion/hooks/base.py [hunk 8]
```diff
@@ -224,7 +224,13 @@ def dispatch(self, *args: Any, **kwargs: Any) -> Any:
...
             return hook.new_forward(self.module, *args, **kwargs)
 
-        # For multiple hooks, chain them in sorted order
+        # For multiple hooks, TeaCache must take precedence if present
+        # because it needs to control whether to skip transformer computation
+        for name, hook in self._hooks.items():
+            if hook.__class__.__name__ == "TeaCacheHook":
+                return hook.new_forward(self.module, *args, **kwargs)
+
+        # For other multiple hooks, chain them in sorted order
```
Comment (lishunyang12) [inline vllm_omni/diffusion/hooks/base.py]: `hook.__class__.__name__ == "TeaCacheHook"` is fragile — breaks with subclassing or renaming. Add a `_takes_dispatch_priority = True` class attribute to `TeaCacheHook` and check that instead.

File: vllm_omni/diffusion/cache/teacache/hook.py [hunk 9]
```diff
@@ -110,6 +110,13 @@ def new_forward(self, module: torch.nn.Module, *args: Any, **kwargs: Any) -> Any
...
         """
+        # Call pre_forward from other hooks (e.g., offload hooks) if registered
+        registry = getattr(module, "_hook_registry", None)
+        if registry is not None:
+            for name, hook in registry._hooks.items():
+                if name != self._HOOK_NAME:
+                    args, kwargs = hook.pre_forward(module, *args, **kwargs)
+
         # Get model-specific context from extractor
         # The extractor encapsulates ALL model-specific logic
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/hook.py:114]: This manually replicates what `dispatch` does for `pre_forward`. If `dispatch` already gives `TeaCacheHook` priority via the `base.py` change, could `TeaCacheHook` call `pre_forward` through the normal chain and do its cache logic in `new_forward`? As-is, any future hook that needs more than `pre_forward` (e.g. its own `post_forward`) is silently skipped.

File: examples/offline_inference/text_to_video/text_to_video.py [hunk 10]
```diff
@@ -240,8 +240,6 @@ def main():
             audio = frames.multimodal_output["audio"]
         if frames.is_pipeline_output and frames.request_output is not None:
             inner_output = frames.request_output
-            if isinstance(inner_output, list):
-                inner_output = inner_output[0] if inner_output else None
             if isinstance(inner_output, OmniRequestOutput):
```
Comment (lishunyang12) [inline examples/offline_inference/text_to_video/text_to_video.py:476]: Why was the `isinstance(inner_output, list)` check removed? Seems unrelated to TeaCache — could this break other models that return a list here?

**Review by lishunyang12 — COMMENTED** [review]
Previous extractor comments look addressed. The offload compat fix needs a couple of tweaks.

File: examples/offline_inference/text_to_video/text_to_video.py (hunk 10 above)
Comment (RuixiangMa) [reply examples/offline_inference/text_to_video/text_to_video.py:476]: This test was modified accidentally，will restore it

File: vllm_omni/diffusion/cache/teacache/hook.py (hunk 9 above)
Comment (RuixiangMa) [reply vllm_omni/diffusion/cache/teacache/hook.py:114]: fixed

**Comment by hsliuustc0106** [comment]
resolve conflicts

**Comment by RuixiangMa** [comment]
> resolve conflicts

fixed
